### PR TITLE
[Compiler Add] allow multiple smart contract exist in one project

### PR DESCRIPTION
**Pull request by Jim8y**
This pr simply enables the compiler to compile multiple smart contracts in one C# project

### CODE REVIEW

File: src/Neo.Compiler.CSharp/CompilationContext.cs [hunk 1]
```diff
@@ -424,40 +312,59 @@ private void ProcessClass(SemanticModel model, INamedTypeSymbol symbol)
             bool isAbstract = symbol.IsAbstract;
             bool isContractType = symbol.IsSubclassOf(nameof(scfx.Neo.SmartContract.Framework.SmartContract));
             bool isSmartContract = isPublic && !isAbstract && isContractType;
+            // If the class is a smart contract, we need to check if it is the expected one
+            // And we can not reuse the compiled methods as their offset is specific for its own contract.
+            // TODO: We can make the smart contract able to call each other directly, but will take a while to implement
```
Comment (shargon) [inline src/Neo.Compiler.CSharp/CompilationContext.cs:317]: I think that is not possible, we need know the hash

File: src/Neo.Compiler.CSharp/Neo.Compiler.CSharp.csproj [hunk 2]
```diff
@@ -19,6 +19,7 @@
 
   <ItemGroup>
     <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="4.8.0" />
+      <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Workspaces" Version="4.8.0" />
```
Comment (shargon) [inline src/Neo.Compiler.CSharp/Neo.Compiler.CSharp.csproj:22]: Is used?

File: src/Neo.Compiler.CSharp/CompilationContext.cs (hunk 1 above)
Comment (Jim8y) [reply src/Neo.Compiler.CSharp/CompilationContext.cs:317]: Debug mode, just specify it manually. An idea in my mind, will try in the future.

File: src/Neo.Compiler.CSharp/Neo.Compiler.CSharp.csproj (hunk 2 above)
Comment (Jim8y) [reply src/Neo.Compiler.CSharp/Neo.Compiler.CSharp.csproj:22]: Should be able to remove it safely, added when I try to figure out an easy way to fetch all smart contracts. Failed obviously. But I am off my pc and is convincing my brain to sleep... so you can do it.

Comment (shargon) [reply src/Neo.Compiler.CSharp/Neo.Compiler.CSharp.csproj:22]: Then, let's wait for tomorrow

**Comment by Jim8y** [comment]
Test engine not support multi-contract. Will need to update the test engine structure. But aside from that, all set.

**Comment by Jim8y** [comment]
@superboyiii can you please do some tests?

**Comment by shargon** [comment]
> Test engine not support multi-contract. Will need to update the test engine structure. But aside from that, all set. Since you are building a new testengine, i think it would be better to update that later in another pr after your test engine is merged if you dont need that function in this pr.

It does, you can call deploy multiple times

File: tests/Neo.SmartContract.TestEngine/TestEngine.cs [hunk 3]
```diff
@@ -60,28 +66,42 @@ public CompilationContext AddNoOptimizeEntryScript(params string[] files)
...
             else
             {
-                context.Diagnostics.ForEach(Console.Error.WriteLine);
+                contexts.ForEach(c => c.Diagnostics.ForEach(Console.Error.WriteLine));
             }
-            return context;
+            return contexts;
         }
 
+
```
Comment (shargon) [inline tests/Neo.SmartContract.TestEngine/TestEngine.cs:104]:
```suggestion
```

Ah you mean the old one, it will be removed after change the unit tests

**Comment by Jim8y** [comment]
> > Test engine not support multi-contract. Will need to update the test engine structure. But aside from that, all set. Since you are building a new testengine, i think it would be better to update that later in another pr after your test engine is merged if you dont need that function in this pr.
> 
> It does, you can call deploy multiple times
> 
> https://github.com/neo-project/neo-devpack-dotnet/blob/4b492d6f2be290cf705bf5ca04720de01d771851/src/Neo.SmartContract.Testing/TestingStandards/Nep17Tests.cs#L192

I mean, specify which one to call.

File: tests/Neo.SmartContract.Framework.TestContracts/Neo.SmartContract.Framework.TestContracts.csproj [hunk 4]
```diff
@@ -16,4 +16,9 @@
       <ProjectReference Include="..\..\src\Neo.SmartContract.Framework\Neo.SmartContract.Framework.csproj" />
     </ItemGroup>
 
+	<Target Name="PreBuild" BeforeTargets="PreBuildEvent">
+		<Message Text="Create and compile Contracts.cs" Importance="high" />
+		<Exec Command="dotnet $(SolutionDir)src/Neo.Compiler.CSharp/$(OutputPath)/nccs.dll $(NullableArgument) $(CheckedArgument) $(DebugArgument) &quot;$(SolutionDir)tests/Neo.SmartContract.Framework.TestContracts/Neo.SmartContract.Framework.TestContracts.csproj&quot; -o &quot;$(SolutionDir)tests/Neo.SmartContract.Framework.UnitTests/Artifacts&quot; --generate-artifacts source" />
```
Comment (shargon) [inline tests/Neo.SmartContract.Framework.TestContracts/Neo.SmartContract.Framework.TestContracts.csproj:21]: It doesn't work @Jim8y

File: src/Neo.Compiler.CSharp/CompilationContext.cs (hunk 1 above)
Comment (shargon) [reply src/Neo.Compiler.CSharp/CompilationContext.cs]: We should fault the compilation until done, I think that is not possible to do it...

File: tests/Neo.SmartContract.Framework.TestContracts/Neo.SmartContract.Framework.TestContracts.csproj (hunk 4 above)
Comment (Jim8y) [reply tests/Neo.SmartContract.Framework.TestContracts/Neo.SmartContract.Framework.TestContracts.csproj:21]: I'll check.

Comment (shargon) [reply tests/Neo.SmartContract.Framework.TestContracts/Neo.SmartContract.Framework.TestContracts.csproj:21]: This should work:
> "xxx\neo-devpack-dotnet\tests/Neo.SmartContract.Framework.TestContracts/Neo.SmartContract.Framework.TestContracts.csproj" -o "xxx\neo-devpack-dotnet\tests/Neo.SmartContract.Framework.UnitTests/Artifacts" --generate-artifacts source

**Comment by Jim8y** [comment]
image
I think I need to check the contract grammars.

**Comment by Jim8y** [comment]
They have conflicts when compiled together under the same syntax tree.

**Comment by Jim8y** [comment]
@Hecate2 can you please help me check this pr? compile works but optimization fail:

run nccs with this command of your own path:

C:\Users\liaoj\git\neo-devpack-dotnet\tests\Neo.SmartContract.Framework.TestContracts\Neo.SmartContract.Framework.TestContracts.csproj -o C:\Users\liaoj\git\neo-devpack-dotnet\tests\Neo.SmartContract.Framework.UnitTests\Artifacts  --generate-artifacts source

**Comment by Hecate2** [comment]
> @Hecate2 can you please help me check this pr? compile works but optimization fail:
> 
> run nccs with this command of your own path:
> 
> C:\Users\liaoj\git\neo-devpack-dotnet\tests\Neo.SmartContract.Framework.TestContracts\Neo.SmartContract.Framework.TestContracts.csproj -o C:\Users\liaoj\git\neo-devpack-dotnet\tests\Neo.SmartContract.Framework.UnitTests\Artifacts  --generate-artifacts source

WILCO

File: src/Neo.Compiler.CSharp/CompilationEngine.cs [hunk 5]
```diff
@@ -98,6 +98,8 @@ public Compilation GetCompilation(string csproj)
             string obj = Path.Combine(folder, "obj");
             HashSet<string> sourceFiles = Directory.EnumerateFiles(folder, "*.cs", SearchOption.AllDirectories)
                 .Where(p => !p.StartsWith(obj))
+                .GroupBy(Path.GetFileName)
```
Comment (shargon) [inline src/Neo.Compiler.CSharp/CompilationEngine.cs:176]: We should test contracts with partial class (multiple files)

**Comment by shargon** [comment]
@Jim8y thanks for do it! It's very important for complex projects! Good work!

File: src/Neo.Compiler.CSharp/CompilationContext.cs [hunk 6]
```diff
@@ -424,40 +306,47 @@ private void ProcessClass(SemanticModel model, INamedTypeSymbol symbol)
...
             bool isContractType = symbol.IsSubclassOf(nameof(scfx.Neo.SmartContract.Framework.SmartContract));
             bool isSmartContract = isPublic && !isAbstract && isContractType;
+
             if (isSmartContract)
             {
-                if (scTypeFound) throw new CompilationException(DiagnosticId.MultiplyContracts, "Only one smart contract is allowed.");
-                scTypeFound = true;
+                if (_targetContract.Name != symbol.Name)
+                {
+                    return;
```
Comment (shargon) [inline src/Neo.Compiler.CSharp/CompilationContext.cs:314]: could you add a comment about when this happend?

File: src/Neo.Compiler.CSharp/CompilationEngine.cs [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+                        break;
+                    default:
+                        throw new NotSupportedException();
+                }
+                MetaReferences.Add(assemblyName, reference);
+            }
+            return reference;
+        }
+
+        private static bool IsSingleAbstractClass(IEnumerable<SyntaxTree> syntaxTrees)
```
Comment (shargon) [inline src/Neo.Compiler.CSharp/CompilationEngine.cs:248]: Does we need to change to check if there are one abstract, not single?

File: tests/Neo.SmartContract.Framework.TestContracts/Neo.SmartContract.Framework.TestContracts.csproj (hunk 4 above)
Comment (shargon) [reply tests/Neo.SmartContract.Framework.TestContracts/Neo.SmartContract.Framework.TestContracts.csproj:21]: still not working to me

**Comment by shargon** [comment]
It seems that is not included in workflow the contract tests

**Comment by shargon** [comment]
@Jim8y conflicts

**Comment by superboyiii** [comment]
> @superboyiii can you please do some tests?

I will

**Comment by shargon** [comment]
@Jim8y It sems that previous file changes appear in changes

File: src/Neo.Compiler.CSharp/CompilationContext.cs (hunk 6 above)
Comment (Jim8y) [reply src/Neo.Compiler.CSharp/CompilationContext.cs:326]: Done

File: src/Neo.Compiler.CSharp/CompilationEngine.cs (hunk 7 above)
Comment (Jim8y) [reply src/Neo.Compiler.CSharp/CompilationEngine.cs]: Done

**Comment by Jim8y** [comment]
this is good to go, please review and test @shargon @superboyiii

File: tests/Neo.Compiler.CSharp.TestContracts/Contract_Multiple.cs [hunk 8]
```diff
@@ -1,7 +1,3 @@
-using System;
-using System.ComponentModel;
-using System.Numerics;
-
 namespace Neo.Compiler.CSharp.UnitTests.TestClasses
 {
     public class Contract_MultipleA : SmartContract.Framework.SmartContract
```
[the remark this answers is not in the document]
Comment (Jim8y) [reply tests/Neo.Compiler.CSharp.TestContracts/Contract_Multiple.cs:3]: [tests/Neo.Compiler.CSharp.TestContracts/Contract_Partial.2.cs](https://github.com/neo-project/neo-devpack-dotnet/blob/7f64dc47077c0e6fcd10fe09a9f80b51737a861c/tests/Neo.Compiler.CSharp.TestContracts/Contract_Partial.2.cs#L4)

**Review by shargon — APPROVED** [review]

**Comment by Jim8y** [comment]
Mine is merged, your turn to fix conflicts now @shargon !!!!!

**Comment by Jim8y** [comment]
Any link? Such that I can check

**Comment by shargon** [comment]
> Any link? Such that I can check

No, just trying to compile something wrong in my laptop

**Pull request merged**